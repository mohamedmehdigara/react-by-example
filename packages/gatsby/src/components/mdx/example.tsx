import React, { FC, ReactNode, useCallback, useState } from 'react'
import styled from 'styled-components'
import { CodeSandboxIcon, ResetIcon } from '../icons'

const Render = styled.div`
  display: flex;
  padding: 1em;
  border: 1px solid var(--app-foreground);
  border-radius: var(--app-border-radius);
  align-items: center;
  justify-content: center;
`
const FlexRight = styled.div`
  display: flex;
  justify-content: flex-end;
`

const CodeSandboxLink = styled.a`
  display: flex;
`

const ResetButton = styled.button`
  display: flex;
  margin: 0;
  padding: 0;
  color: var(--app-foreground);
  background: none;
  border: none;
  cursor: pointer;

  :focus,
  :active {
    color: var(--app-cyan);
    outline: 2px solid var(--app-cyan);
  }
`

export interface ExampleProps {
  code?: ReactNode
  render?: ReactNode
  link?: string
}

export const Example: FC<ExampleProps> = ({ code, render, link }) => {
  const [resetKey, setResetKey] = useState(Date.now())
  const triggerReset = useCallback(() => {
    setResetKey(Date.now())
  }, [])
  return (
    <>
      {code}
      <Render key={resetKey}>
        <div>{render}</div>
      </Render>
      {link && (
        <FlexRight>
          <ResetButton
            type="button"
            title="reset example"
            onClick={triggerReset}
          >
            <ResetIcon />
          </ResetButton>
          <CodeSandboxLink
            target="_blank"
            rel="noreferrer"
            href={link}
            title="edit on CodeSandbox"
          >
            <CodeSandboxIcon />
          </CodeSandboxLink>
        </FlexRight>
      )}
    </>
  )
}
