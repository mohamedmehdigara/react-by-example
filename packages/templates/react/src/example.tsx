import React, { FC, ReactElement } from 'react'

export const Simple: FC = () => <h1>Hello world!</h1>

export default (): ReactElement => <Simple />
