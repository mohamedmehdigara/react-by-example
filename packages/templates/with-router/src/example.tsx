import React, { FC } from 'react'

export const Simple: FC = () => <h1>Hello world!</h1>

export default <Simple />
