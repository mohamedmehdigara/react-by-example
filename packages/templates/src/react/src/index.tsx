import React from 'react'
import ReactDOM from 'react-dom'
import example from './example'

ReactDOM.render(<>{example}</>, document.getElementById('root'))
