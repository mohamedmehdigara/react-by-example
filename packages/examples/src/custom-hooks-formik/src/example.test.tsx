import { act, cleanup, render, RenderResult } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React, { FC, useContext } from 'react'
import example, {
  DebugForm,
  Form,
  FormContext,
  Input,
  Select,
  useForm,
} from './example'

describe('DebugForm', () => {
  let component: RenderResult

  beforeEach(() => {
    component = render(
      <Form initialValues={{ hello: 'world' }}>
        <DebugForm />
      </Form>,
    )
  })

  afterEach(() => {
    cleanup()
  })

  it('should snapshot', () => {
    expect(component.container).toMatchSnapshot()
  })
})

describe('Form', () => {
  let stateSpy: jest.Mock
  let setStateSpy: jest.Mock

  const MockFormItem: FC = () => {
    const form = useContext(FormContext)

    stateSpy(form.state)
    setStateSpy.mockImplementation(form.setState)

    return null
  }

  beforeEach(() => {
    stateSpy = jest.fn()
    setStateSpy = jest.fn()

    render(
      <Form initialValues={{ hello: 'world' }}>
        <MockFormItem />
      </Form>,
    )
  })

  afterEach(() => {
    cleanup()
  })

  it('should set initial state to initialValues', () => {
    expect(stateSpy).toHaveBeenLastCalledWith({ hello: 'world' })
    expect(stateSpy).toHaveBeenCalledTimes(1)
  })

  describe('when updating state', () => {
    beforeEach(() => {
      act(() => {
        setStateSpy({ hello: 'new state' })
      })
    })

    it('should update state', () => {
      expect(stateSpy).toHaveBeenLastCalledWith({ hello: 'new state' })
    })

    describe('when extending state', () => {
      beforeEach(() => {
        act(() => {
          setStateSpy({ goodbye: 'world' })
        })
      })

      it('should merge state', () => {
        expect(stateSpy).toHaveBeenLastCalledWith({
          hello: 'new state',
          goodbye: 'world',
        })
      })
    })
  })
})

describe('useForm', () => {
  let component: RenderResult
  let textarea: HTMLElement

  const TextArea: FC = () => {
    return <textarea {...useForm('textarea')} />
  }

  beforeEach(() => {
    component = render(
      <Form initialValues={{ textarea: 'initial' }}>
        <TextArea />
        <DebugForm />
      </Form>,
    )
    textarea = component.getByRole('textbox')
  })

  afterEach(() => {
    cleanup()
  })

  it('should snapshot', () => {
    expect(component.container).toMatchSnapshot()
  })

  it('should use initial value', () => {
    expect(textarea).toHaveValue('initial')
  })

  describe('when typing', () => {
    beforeEach(async () => {
      await userEvent.type(textarea, '{selectall}new value')
    })

    it('should update value', () => {
      expect(textarea).toHaveValue('new value')
      expect(component.container).toHaveTextContent(/"textarea": "new value"/)
    })
  })
})

describe('Input', () => {
  let component: RenderResult
  let input: HTMLElement

  beforeEach(() => {
    component = render(
      <Form initialValues={{ input: '' }}>
        <Input name="input" />
        <DebugForm />
      </Form>,
    )
    input = component.getByRole('textbox')
  })

  afterEach(() => {
    cleanup()
  })

  it('should snapshot', () => {
    expect(component.container).toMatchSnapshot()
  })

  it('should use initial value', () => {
    expect(input).toHaveValue('')
  })

  describe('when typing', () => {
    beforeEach(async () => {
      await userEvent.type(input, '{selectall}updated value')
    })

    it('should update value', () => {
      expect(input).toHaveValue('updated value')
      expect(component.container).toHaveTextContent(/"input": "updated value"/)
    })
  })
})

describe('Select', () => {
  let component: RenderResult
  let select: HTMLElement

  beforeEach(() => {
    component = render(
      <Form initialValues={{ select: 'red' }}>
        <Select name="select">
          <option>red</option>
          <option>green</option>
          <option>blue</option>
        </Select>
        <DebugForm />
      </Form>,
    )
    select = component.getByRole('combobox')
  })

  afterEach(() => {
    cleanup()
  })

  it('should snapshot', () => {
    expect(component.container).toMatchSnapshot()
  })

  it('should use initial value', () => {
    expect(select).toHaveValue('red')
  })

  describe('when choosing', () => {
    beforeEach(() => {
      userEvent.selectOptions(select, 'blue')
    })

    it('should update value', () => {
      expect(select).toHaveValue('blue')
      expect(component.container).toHaveTextContent(/"select": "blue"/)
    })
  })
})

describe('example', () => {
  let component: RenderResult

  beforeEach(() => {
    component = render(<>{example}</>)
  })

  afterEach(() => {
    cleanup()
  })

  it('should snapshot', () => {
    expect(component.container).toMatchSnapshot()
  })
})
