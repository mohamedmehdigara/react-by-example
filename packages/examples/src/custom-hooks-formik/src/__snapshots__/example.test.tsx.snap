// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DebugForm should snapshot 1`] = `
<div>
  <form>
    <pre>
      {
  "hello": "world"
}
    </pre>
  </form>
</div>
`;

exports[`Input should snapshot 1`] = `
<div>
  <form>
    <input
      value=""
    />
    <pre>
      {
  "input": ""
}
    </pre>
  </form>
</div>
`;

exports[`Select should snapshot 1`] = `
<div>
  <form>
    <select>
      <option>
        red
      </option>
      <option>
        green
      </option>
      <option>
        blue
      </option>
    </select>
    <pre>
      {
  "select": "red"
}
    </pre>
  </form>
</div>
`;

exports[`example should snapshot 1`] = `
<div>
  <form>
    <label>
      title
      <select>
        <option
          value="mr"
        >
          Mr
        </option>
        <option
          value="mrs"
        >
          Mrs
        </option>
        <option
          value="ms"
        >
          Ms
        </option>
        <option
          value="dr"
        >
          Dr
        </option>
        <option
          value="lord"
        >
          Lord
        </option>
        <option
          value="prof"
        >
          Prof
        </option>
        <option
          value="esq"
        >
          Esq
        </option>
        <option
          value="dame"
        >
          Dame
        </option>
        <option
          value="na"
        >
          N/A
        </option>
      </select>
    </label>
    <br />
    <label>
      first name
      <input
        value="Steve"
      />
    </label>
    <br />
    <label>
      last name
      <input
        value="Buscemi"
      />
    </label>
    <pre>
      {
  "title": "mr",
  "first": "Steve",
  "last": "Buscemi"
}
    </pre>
  </form>
</div>
`;

exports[`useForm should snapshot 1`] = `
<div>
  <form>
    <textarea>
      initial
    </textarea>
    <pre>
      {
  "textarea": "initial"
}
    </pre>
  </form>
</div>
`;
